### DISCUSSION TITLE
ACBF Editor - My thinking

### DISCUSSION THREAD

**Discussion by mizaki** [GeoRW/ACBF-Editor#14] [General]
This is my thinking on what the ACBF Editor should do.

This [BallonsTrandlator](https://github.com/dmMaze/BallonsTranslatorproject) is good and I don't think it makes sense to try and replicate a lot of what it does with the ACBF Editor. I plan to approach BT to see if they would be interested in saving to ACBF format.

With that said, having the basics of frame and bubble detection in the editor makes sense. If that needs to be "AI" for the bubbles, I don't know if going that far is required? I will also mention that frame detection is still pretty bad and I've not found anything other than the kumiko approach. There is this: https://github.com/pedrovgs/DeepPanel but if I'm reading it right, we'd need to train it ourselves (or somehow use the Android model).

My main reason for wanting to revive the ACBF format is for guided reading and accessibility (and being able to highlight text for Japanese reading). I'll put my feature requests in the Viewer repo though.

Any thoughts on direction?

**Comment by GeoRW** [comment]
I would keep at least basic stuff in ACBF Editor the way it is now, so one can work with ACBF Editor itself.

If for frame detection improvement we want to add just one library, then that I would do, but maybe rather add it directly to ACBF Editor code rather than link as external dependency to install. Or maybe have current frame detection code there as default and optionally use external library in case it is available/installed.

Bubble detection algorithm works on single bubbles pretty well, drawing more complex bubbles could be improved by adding option to draw ellipses. Or we can also improve detection algorithm and when it detects joined bubbles, it creates polygon for each.

Once bubble boundaries are defined, shouldn't be a big problem to run the image through OCR to read the text there. Maybe this can be added as optional dependency too (most distributions have some OCR libraries installed by default I believe, not sure about python bindings). This was one of the things I wanted to do in the past, but didn't get to it. Once text is there, adding another library to translate it could be a next step.

I don't know if that Balloons Translator is detecting bubble boundaries, it seems pretty complex editor where user deletes the old text in underlying image and then selects the font size, style himself ad resizes it to fit it into the bubble. Then it exports resulting image, I believe. ACBF is based on bubble polygon and fits text inside automatically. Balloons Translator is also dependent on some online resources, trained model and yes that's too much to add into ACBF Editor itself and I would rather avoid these kind of dependencies. Also not sure if they would easily add support for export into ACBF schema. Does it even work on multiple images at ones?

If it's good, then maybe I (as a translator) would use it side by side with ACBF Editor and just copy text and translations over, if that would speed up the workflow. Normally, I can translate most of the stuff in my head and have google translate opened in other window to help me find better words/expressions if I need it.

**Comment by GeoRW** [comment]
Also regarding drawing in ACBF Editor, I made it quick and dirty to draw just points with one mouse click. That can be improved to draw different objects and use click-and-hold then release ... So lot of work to do.
Don't know how complex it is for you to port everything to GTK4. Didn't have a chance to upgrade my machine yet.

**Comment by mizaki** [comment]
I agree that being able to do the basics is required, it's how far outside of that to go if other projects already do the same thing.

Ballons can automatically translate whole comics. Give it a try, there are offline options for most. It also saves a JSON file (without frames last time I looked, so that needs to be requested) which being able to import was my other option.

Using outside libraries is a double edged sword but it does reduce work. Here are some other options (not against homegrown and these could be optional).
(Old but could be forked) https://github.com/AsiriAmalk/Comic_Speech_Bubble_Identifier_Using_Pytesseract/tree/master
(Old) https://github.com/damishshah/comic-book-reader
(Used by Ballons) https://github.com/dmMaze/comic-text-detector
https://github.com/zyddnys/manga-image-translator
https://github.com/KUR-creative/SickZil-Machine

These may not all be useful.

There are already external requirements and even if they aren't used, it would still pay to "externalize" these functions to a python module. With something like kumiko, there's no reason not to do PRs for the protect to help improve it and if required, fork it for our own use.

For drawing tools, really all that's needed is lines and boxes?

**Reply by GeoRW** [reply]
Ok, if Baloons can produce usable output to import into ACBF Editor, then we can do something like "import text-layer from a file ..."
Regarding drawing tools, what I miss at the moment is to draw ellipse. I will add that to GTK3 version (should be easy to do from 3 points). But the whole drawing could be done better.
So, for drawing (in case bubble detection doesn't work properly or you need overlay some non-bubble text), it is useful to have:
- polygons
- rectangles
- ellipses
- rectangle with rounded corners (this is nicer than a rectangle and I use it for translating titles/headings)
- maybe something for sounds, like a jagged ellipse
